storage: avoid double lock in tryGetOrCreateReplica

nvb opened the pull request:
Before this change, `tryGetOrCreateReplica` was RLocking, checking
an uncommon condition, then Locking. This was unnecessary and made
the code look strange. It also resulted in an extra RLock in the
exceptionally common case where the Replica was not destroyed.

This change fixes this by collapsing the destroyStatus check into
the main critical section. There's no real expected performance
improvement here, this is mostly just for cleaner code.

Release note: None

nvb committed:
storage: avoid double lock in tryGetOrCreateReplica

Before this change, `tryGetOrCreateReplica` was RLocking, checking
an uncommon condition, then Locking. This was unnecessary and made
the code look strange. It also resulted in an extra RLock in the
exceptionally common case where the Replica was not destroyed.

This change fixes this by collapsing the destroyStatus check into
the main critical section. There's no real expected performance
improvement here, this is mostly just for cleaner code.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/28002)
<!-- Reviewable:end -->

nvb committed:
storage: avoid another RLock in tryGetOrCreateReplica

This avoids another read lock in the common path of
tryGetOrCreateReplica. The lock was being acquired
in `Replica.Desc`.

Release note: None

nvb committed:
storage: acquire read lock in Replica.requestCanProceed

This didn't need to acquire a write lock.

Release note: None

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28002#-:-LIXB6JHDo3TBFLp4Eqh:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
TFTR!

bors r+

nvanbenschoten commented:
Flake was https://github.com/cockroachdb/cockroach/issues/28004.

bors r+
